### PR TITLE
Male inty

### PR DESCRIPTION
**Pull request by tkstanczak**

### CODE REVIEW

File: src/Nethermind/Nethermind.Cli/Modules/EthCliModule.cs [hunk 1]
```diff
@@ -83,7 +83,7 @@ public JsValue GetBlockByNumber(string blockParameter, bool returnFullTransactio
         [CliFunction("eth", "sendEth")]
         public string SendEth(string from, string to, decimal amountInEth)
         {
-            return SendEth(CliParseAddress(from), CliParseAddress(to), (UInt256) (amountInEth * (decimal) 1.Ether()));
+            return SendEth(CliParseAddress(from), CliParseAddress(to), (UInt256) (amountInEth * (decimal)(BigInteger) 1.Ether()));
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Cli/Modules/EthCliModule.cs:86]: again whats the point with BigInteger?

File: src/Nethermind/Nethermind.Clique.Test/CliqueBlockProducerTests.cs [hunk 2]
```diff
@@ -404,7 +404,7 @@ public On AddPendingTransaction(PrivateKey nodeKey)
                 transaction.To = TestItem.AddressC;
                 transaction.GasLimit = 30000;
                 transaction.GasPrice = 20.GWei();
-                transaction.Nonce = _currentNonce++;
+                transaction.Nonce = _currentNonce + 1;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Clique.Test/CliqueBlockProducerTests.cs:407]: looks like potential issue to me, _currentNonce is not being modified.

File: src/Nethermind/Nethermind.Consensus.AuRa/AuraDifficultyCalculator.cs [hunk 3]
```diff
@@ -24,10 +24,20 @@ public static class AuraDifficultyCalculator
 
         static AuraDifficultyCalculator()
         {
-            UInt256.Create(out MaxDifficulty, UInt128.MaxValue, UInt128.Zero);
+            MaxDifficulty = new UInt256(ulong.MaxValue, ulong.MaxValue, 0, 0);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus.AuRa/AuraDifficultyCalculator.cs:27]: Make UInt128MaxValue a constant

File: src/Nethermind/Nethermind.Consensus.Ethash/DifficultyCalculator.cs [hunk 4]
```diff
@@ -39,12 +39,12 @@ public DifficultyCalculator(ISpecProvider specProvider)
             bool parentHasUncles)
         {
             IReleaseSpec spec = _specProvider.GetSpec(blockNumber);
-            BigInteger baseIncrease = BigInteger.Divide(parentDifficulty, spec.DifficultyBoundDivisor);
-            BigInteger timeAdjustment = TimeAdjustment(spec, parentTimestamp, currentTimestamp, parentHasUncles);
+            BigInteger baseIncrease = BigInteger.Divide((BigInteger)parentDifficulty, spec.DifficultyBoundDivisor);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus.Ethash/DifficultyCalculator.cs:42]: still BigInteger?

File: src/Nethermind/Nethermind.Consensus.Ethash/Ethash.cs [hunk 5]
```diff
@@ -168,7 +168,7 @@ private bool IsLessThanTarget(byte[] result, BigInteger target)
 
             ulong fullSize = GetDataSize(epoch);
             ulong nonce = startNonce ?? GetRandomNonce();
-            BigInteger target = BigInteger.Divide(_2To256, header.Difficulty);
+            BigInteger target = BigInteger.Divide(_2To256, (BigInteger)header.Difficulty);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus.Ethash/Ethash.cs:171]: BigInteger?

File: src/Nethermind/Nethermind.Evm/EvmStack.cs [hunk 6]
```diff
@@ -168,39 +167,9 @@ public void PushUInt256(ref UInt256 value)
...
-            fullBytes.Slice(start, length).CopyTo(fillCount > 0 ? word.Slice(fillCount, 32 - fillCount) : word);
-
-            if (_tracer.IsTracingInstructions) _tracer.ReportStackPush(word);
-
-            if (++Head >= MaxStackSize)
-            {
-                Metrics.EvmExceptions++;
-                throw new EvmStackOverflowException();
-            }
+            PushUInt256((UInt256) value);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/EvmStack.cs:172]: Not sure whats going on here

File: src/Nethermind/Nethermind.Serialization.Rlp/RlpStream.cs [hunk 7]
```diff
@@ -358,7 +358,7 @@ public void Encode(long value)
 
         public void Encode(ulong value)
         {
-            Encode(value, 8);
+            Encode((UInt256)value, 8);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Serialization.Rlp/RlpStream.cs:361]: why do we need this cast?

File: src/Nethermind/Nethermind.BeamWallet/Modules/Transfer/TransferModule.cs [hunk 8]
```diff
@@ -199,13 +202,13 @@ private async Task GetEstimateGasAsync()
 
         private void SetEstimateGas(decimal estimateGasResult)
         {
-            if (estimateGasResult > (decimal) _gasLimit)
+            if (estimateGasResult > (decimal) (BigInteger) _gasLimit)
```
[the remark this answers is not in the document]
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.BeamWallet/Modules/Transfer/TransferModule.cs:205]: Can we port decimal cast & create from dirchlet?

public static explicit operator decimal(UInt256 a)
        {
            if (a.s1 == 0)
                return a.s0;
            var shift = Math.Max(0, 32 - GetBitLength(a.s1));
            UInt256 ashift;
            RightShift(out ashift, ref a, shift);
            return new decimal((int) a.r0, (int) a.r1, (int) a.r2, false, (byte) shift);
        }

File: src/Nethermind/Nethermind.Evm/Precompiles/ModExpPrecompile.cs [hunk 9]
```diff
@@ -60,7 +60,7 @@ public long DataGasCost(byte[] inputData, IReleaseSpec releaseSpec)
                 UInt256 lengthOver32 = expLength <= 32 ? 0 : expLength - 32;
                 UInt256 adjusted = AdjustedExponentLength(lengthOver32, expSignificantBytes);
                 UInt256 gas = complexity * UInt256.Max(adjusted, UInt256.One) / 20;
-                return (long)gas;
+                return gas > long.MaxValue ? long.MaxValue : (long)gas;
```
[the remark this answers is not in the document]
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Evm/Precompiles/ModExpPrecompile.cs:63]: hmmm?

File: src/Nethermind/Nethermind.BeamWallet/Modules/Transfer/TransferModule.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (tkstanczak) [reply src/Nethermind/Nethermind.BeamWallet/Modules/Transfer/TransferModule.cs:205]: hmmm, it was ported partially

File: src/Nethermind/Nethermind.Blockchain.Test/TestTransactionsGenerator.cs [hunk 10]
```diff
@@ -76,7 +76,7 @@ private void TimerOnElapsed(object sender, ElapsedEventArgs elapsedEventArgs)
             tx.GasPrice = 1;
             tx.GasLimit = 21000;
             tx.To = new Address(0x0f.ToBigEndianByteArray().PadLeft(20));
-            tx.Nonce = _nonce++;
+            tx.Nonce = _nonce + 1;
```
[the remark this answers is not in the document]
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Blockchain.Test/TestTransactionsGenerator.cs:79]: good catch, I was tired with lack of operators

[the remark this answers is not in the document]
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Blockchain.Test/TestTransactionsGenerator.cs]: the class was not used, removed it

File: src/Nethermind/Nethermind.Consensus.Ethash/DifficultyCalculator.cs (hunk 4 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Consensus.Ethash/DifficultyCalculator.cs:42]: yeah, too much hassle here with negative values that need to be Int256 and then applied in operation sto UInt256 - lots of conversions / operators are missing so leaving it here

File: src/Nethermind/Nethermind.Serialization.Rlp/RlpStream.cs (hunk 7 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Serialization.Rlp/RlpStream.cs:361]: ambiguous between UInt256 and BigInteger, BigInteger needed for negative values when encoding long

File: src/Nethermind/Nethermind.Evm/EvmStack.cs (hunk 6 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Evm/EvmStack.cs:172]: replacing PushBigInteger with the new code - diff tool confused you

File: src/Nethermind/Nethermind.Consensus.Ethash/Ethash.cs (hunk 5 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Consensus.Ethash/Ethash.cs]: _2To256 is out of range of UInt256

File: src/Nethermind/Nethermind.Cli/Modules/EthCliModule.cs (hunk 1 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Cli/Modules/EthCliModule.cs]: fixed

File: src/Nethermind/Nethermind.Consensus.AuRa/AuraDifficultyCalculator.cs (hunk 3 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Consensus.AuRa/AuraDifficultyCalculator.cs]: done

File: src/Nethermind/Nethermind.BeamWallet/Modules/Transfer/TransferModule.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (tkstanczak) [reply src/Nethermind/Nethermind.BeamWallet/Modules/Transfer/TransferModule.cs]: it was ported, just initially I did BigInteger cast before I proted it and later the code stayed, fixed

File: src/Nethermind/Nethermind.AuRa.Test/AuraDifficultyCalculatorTests.cs [hunk 11]
```diff
@@ -32,20 +32,20 @@ private static IEnumerable DifficultyTestCases
...
-                yield return new TestCaseData(10, 0, 5).Returns(new UInt256(UInt128.MaxValue - 5));
-                yield return new TestCaseData(10, 9, 3).Returns(new UInt256(UInt128.MaxValue) + 2);
-                yield return new TestCaseData(100, 10, 10).Returns(new UInt256(UInt128.MaxValue - 80));
+                yield return new TestCaseData(1, 0, 0).Returns(new UInt256(ulong.MaxValue - 1, ulong.MaxValue, 0, 0));
+                yield return new TestCaseData(10, 0, 0).Returns(new UInt256(ulong.MaxValue - 10, ulong.MaxValue, 0, 0));
+                yield return new TestCaseData(10, 9, 0).Returns(new UInt256(ulong.MaxValue - 1, ulong.MaxValue, 0, 0));
+                yield return new TestCaseData(100, 10, 0).Returns(new UInt256(ulong.MaxValue - 90, ulong.MaxValue, 0, 0));
+
+                yield return new TestCaseData(1, 0, 1).Returns(new UInt256(ulong.MaxValue, ulong.MaxValue, 0, 0));
+                yield return new TestCaseData(10, 0, 5).Returns(new UInt256(ulong.MaxValue - 5, ulong.MaxValue, 0, 0));
```
[the remark this answers is not in the document]
Comment (tkstanczak) [reply src/Nethermind/Nethermind.AuRa.Test/AuraDifficultyCalculatorTests.cs]: done

**Pull request closed without merging**